envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt

```diff
@@ -90,6 +94,50 @@ class SnapshotDebugController(controlPlane: ControlPlane) {
         }
     }
 
+    @GetMapping("/snapshot-global/{service}")
+    fun serviceSnapshot(
+        @PathVariable service: String,
+        @RequestParam dc: String?,
+        @RequestParam(defaultValue = "false") xds: Boolean
+    ): ResponseEntity<EndpointInfoList> {
+        val updateResult = snapshotUpdater.getGlobalSnapshot()
+
+        val globalSnapshot = if (xds) {
+            updateResult?.xdsSnapshot
+        } else {
+            updateResult?.adsSnapshot
+        }
+
+        val endpoints = extractEndpoints(globalSnapshot, service)
+
+        if (endpoints == null) {
+            return ResponseEntity(HttpStatus.NOT_FOUND)
+        }
+
+        val clusterInfos = endpoints.endpointsList
+            .filter { dc == null || it.locality.zone.endsWith(dc, true) }
+            .flatMap { locality ->
+                locality.lbEndpointsList.map {
+                    val socketAddress = it.endpoint.address.socketAddress
+                    EndpointInfo(locality.locality.zone, socketAddress.address, socketAddress.portValue)
+                }
+            }
+        return ResponseEntity(EndpointInfoList(clusterInfos), HttpStatus.OK)
+    }
+
+    private fun extractEndpoints(globalSnapshot: GlobalSnapshot?, service: String): ClusterLoadAssignment? {
+        if (globalSnapshot == null) {
+            logger.warn("Global snapshot is missing")
+            return null
+        }
+        val endpoints = globalSnapshot.endpoints
+            .resources()[service]
+        if (endpoints == null) {
+            logger.warn("Can not find $service in global snapshot")
```

Comment:
w metodzie wyzej srpawdzasz juz czy endpoints to null - usunałbym i przeniosl logowanie tam